flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java

```diff
@@ -207,11 +210,28 @@ else if (bufferOrEvent.getEvent().getClass() == CancelCheckpointMarker.class) {
 				if (bufferOrEvent.getEvent().getClass() == EndOfPartitionEvent.class) {
 					processEndOfPartition();
 				}
-				return bufferOrEvent;
+				return next;
 			}
 		}
 	}
 
+	private Optional<BufferOrEvent> handleEmptyBuffer() throws Exception {
+		if (!inputGate.isFinished()) {
+			return Optional.empty();
+		}
+
+		if (!endOfStream) {
```

Comment:
Does not follow your advice to put the simple case first and even has to negate the predicate to do so. You could move the else branch first and out.